src/accessories/characteristics/targetPosition.ts

```diff
@@ -48,50 +44,59 @@ export class TargetPositionCharacteristic extends TuyaWebCharacteristic {
 
-    const data = { state: value === 0 ? 3 : 1 };
+    const coverAccessory = <CoverAccessory>this.accessory;
+    const target = value ? 100 : 0;
+
+    this.debug("Setting targetPosition to %d", target);
 
     this.accessory
-      .setDeviceState("turnOnOff", { value }, data)
+      .setDeviceState("turnOnOff", { value }, value)
       .then(async () => {
-        this.debug("[SET] %s", value);
+        this.debug("[SET] turnOnOff command sent with value %s", value);
         callback();
 
-        await delay(1000);
-        this.accessory.setTuyaCharacteristic(
+        this.debug("Setting targetPosition to %d", target);
+        coverAccessory.target = target;
+        this.accessory.setCharacteristic(
+          this.accessory.platform.Characteristic.TargetPosition,
+          target,
+          true
+        );
+
+        coverAccessory.motor = value
+          ? this.accessory.platform.Characteristic.PositionState.INCREASING
+          : this.accessory.platform.Characteristic.PositionState.DECREASING;
+        this.accessory.setCharacteristic(
+          this.accessory.platform.Characteristic.PositionState,
+          coverAccessory.motor,
+          true
+        );
+
+        await delay(5000);
+
+        this.debug(
+          "Setting currentPosition to %d and positionState to STOPPED",
+          target
+        );
+
+        coverAccessory.position = target;
+        this.accessory.setCharacteristic(
           this.accessory.platform.Characteristic.CurrentPosition,
-          data
+          coverAccessory.position,
+          true
+        );
+
+        coverAccessory.motor = this.accessory.platform.Characteristic.PositionState.STOPPED;
+        this.accessory.setCharacteristic(
+          this.accessory.platform.Characteristic.PositionState,
+          this.accessory.platform.Characteristic.PositionState.STOPPED,
+          true
         );
       })
       .catch(this.accessory.handleError("SET", callback));
   }
 
   updateValue(data: DeviceState, callback?: CharacteristicGetCallback): void {
-    if (!isNaN(Number(String(data?.state)))) {
-      //State is a number and probably 1, 2 or 3
-      const state = Number(data.state);
-      const stateValue = {
-        1: 100,
-        2: 50,
-        3: 0,
-      }[state];
-
-      this.accessory.setCharacteristic(
-        this.homekitCharacteristic,
-        stateValue,
-        !callback
-      );
-      callback && callback(null, stateValue);
-    } else if (["true", "false"].includes(String(data?.state).toLowerCase())) {
-      const stateValue = String(data.state).toLowerCase() === "true" ? 100 : 0;
-      this.accessory.setCharacteristic(
-        this.homekitCharacteristic,
-        stateValue,
-        !callback
-      );
-      callback && callback(null, stateValue);
-    } else {
-      callback &&
-        callback(new Error(`Unexpected state value provided: ${data?.state}`));
-    }
+    callback && callback(null, (<CoverAccessory>this.accessory).target);
```

Comment:
This will fail when `this.accessory` is an instance of `WindowAccessory` which also uses this characteristic